internal/muxado/buffer.go

```diff
@@ -57,9 +60,47 @@ DONE:
 	return n, err
 }
 
+// Notify readers that the deadline has arrived.
+func (b *inboundBuffer) notifyDeadline() {
+	// It's important that the mutex is locked for this. It ensures that an
+	// in-flight timer can't broadcast before a reader gets to its condvar.Wait().
+	b.mu.Lock()
+	b.cond.Broadcast()
+	b.mu.Unlock()
+}
+
+// Start or reset the timer.
+// Must be called when the mutex is locked.
+func (b *inboundBuffer) startTimerLocked(timeout time.Duration) {
+	if b.timer == nil {
+		b.timer = time.AfterFunc(timeout, b.notifyDeadline)
+	} else {
+		b.timer.Reset(timeout)
+	}
+}
+
+// Stops a timer, if one is set.
+// Must be called while the mutex is locked.
+func (b *inboundBuffer) stopTimerLocked() {
+	if b.timer != nil {
+		b.timer.Stop()
+	}
+}
+
 func (b *inboundBuffer) Read(p []byte) (n int, err error) {
 	b.mu.Lock()
 	for {
+		// If the deadline is set, we need to take it into account
+		if !b.deadline.IsZero() {
+			// If the deadline is in the past, bail out.
+			// SetDeadline will ensure that we get woken back up if it expires.
+			if time.Until(b.deadline) < 0 {
+				n = 0
```

Comment:
technically, no need for `n = 0`, right?

Reply:
I never fully trust these named return values, so I like being explicit just on the off-chance that someone later breaks these assumptions today :stuck